test/verify/check-apps

```diff
@@ -0,0 +1,150 @@
+
+    def testBasic(self):
+        b = self.browser
+        m = self.machine
+
+        # Make sure none of the appstream directories exist.  They
+        # will be created later and we need to cope with that.
+
+        m.execute("rm -rf /usr/share/metainfo /usr/share/app-info /var/cache/app-info")
+
+        m.start_cockpit()
+        b.login_and_go("/apps")
+        b.wait_present(".app-list-empty")
+
+        self.createAppStreamPackage("app-1", "1.0", "1")
+        self.enableAppStreamRepo()
+
+        b.wait_present(".app-list tr:contains('app-1') button:contains('Install')")
+        b.click(".app-list tr:contains('app-1') button:contains('Install')")
+
+        b.wait_present(".app-list tr:contains('app-1') button:contains('Remove')")
+        b.click(".app-list tr:contains('app-1') button:contains('Remove')")
+
+        b.wait_present(".app-list tr:contains('app-1') button:contains('Install')")
+
+    def testL10N(self):
+        b = self.browser
+        m = self.machine
+
+        # Switching to a language might produce these messages, which seem to be harmless.
+        self.allow_journal_messages("invalid or unusable locale.*",
+                                    "Error receiving data: Connection reset by peer")
+
+        m.start_cockpit()
+        b.login_and_go("/apps")
+        b.wait_present(".app-list-empty")
+
+        m.execute("mkdir -p /usr/share/app-info/xmls")
+        m.write("/usr/share/app-info/xmls/test.xml", """
+<components origin="test">
+  <component type="addon">
+    <extends>cockpit.desktop</extends>
+    <id>org.cockpit-project.foo</id>
+    <name>NAME:none</name>
+    <name xml:lang="de">NAME:de</name>
+    <summary>SUMMARY:none</summary>
+    <summary xml:lang="de">SUMMARY:de</summary>
+    <description>
+        <p>DESCRIPTION:none</p>
+        <p xml:lang="de">DESCRIPTION:de</p>
+    </description>
+    <launchable type="cockpit-manifest">foo</launchable>
+    <pkgname>foo</pkgname>
+  </component>
+</components>""")
+
+        b.wait_present(".app-list tr:contains('SUMMARY:none') button:contains('Install')")
+        b.click(".app-list tr:contains('SUMMARY:none')")
+
+        b.wait_present(".app-description:contains('DESCRIPTION:none')")
+
+        def set_lang(lang):
+            b.switch_to_top()
+            b.click("#content-user-name")
+            b.click(".display-language-menu a")
+            b.wait_popup('display-language')
+            b.set_val("#display-language select", lang)
+            b.click("#display-language-select-button")
+            b.expect_load()
+            b.wait_present("#content")
+            b.enter_page("/apps")
+
+        set_lang("de-de")
+        b.wait_present(".app-description:contains('DESCRIPTION:de')")
+        b.wait_not_present(".app-description:contains('DESCRIPTION:none')")
+
+        set_lang("ja-ja")
+        b.wait_present(".app-description:contains('DESCRIPTION:none')")
+        b.wait_not_present(".app-description:contains('DESCRIPTION:de')")
+
```

Comment:
Would be nice to have a test case that covers error handling, e. g. with absent packages, broken XML, etc.

Reply:
Yes.  Doing this immediately pointed out some bugs.